idseq_dag/util/s3.py

```diff
@@ -102,6 +103,36 @@ def install_s3mi(installed={}, mutex=TraceLock("install_s3mi", threading.RLock()
             installed['time'] = time.time()
 
 
+credentials_cached = {}
+
+
+def cache_AWS_credentials(max_age_seconds=600):
+    """
+    On EC2/ECS, the AWS CLI calls the instance/container metadata service to fetch instance profile/role credentials
+    every time it runs, unless it finds credentials in environment variables or config files.
+
+    There is a rate limit on the metadata service, and we trigger it sometimes with the following error:
+
+        Error when retrieving credentials from container-role: Error retrieving metadata: Received non 200 response
+        (429) from ECS metadata: You have reached maximum request limit.
+
+    To avoid this error, we fetch the credentials using standard botocore logic and cache them in environment variables.
+
+    These credentials are short-lived, so we refresh them if we last cached them more than 10 minutes ago.
+    """
+    if credentials_cached or "AWS_ACCESS_KEY_ID" not in os.environ:
+        if time.time() - credentials_cached.get("AWS", 0) > max_age_seconds:
+            for var in "AWS_ACCESS_KEY_ID", "AWS_SECRET_ACCESS_KEY", "AWS_SESSION_TOKEN", "AWS_DEFAULT_REGION":
+                os.environ.pop(var, None)
+            session = botocore.session.Session()
+            credentials = session.get_credentials()
+            os.environ["AWS_ACCESS_KEY_ID"] = credentials.access_key
```

Comment:
As commented offline, some batches can take longer than 1 hour to execute, and I'm afraid that setting environment variables in a stateful way like this could cause future AWS commands not related to `fetch_from_s3` fail when these credential in the environment variables get expired.

Not sure if that is the case, but this stateful logic using environment variables would make it even harder to reason about and trace, specially because it would only happen with batches that were taking hours to execute.

My suggestion is instead of setting them with `os.environment`, we could set these environment variables for each command in a stateless, something like:
```python3
    envs = os.environ.copy()
    envs.update(cached_aws_credentials())
    subprocess.open(..., env=envs,...)
```
This can be easily changed on `command_patterns.py` [here](https://github.com/chanzuckerberg/idseq-dag/blob/b96a29efe73c6bc82ea7dbc8debb352f3a62dce0/idseq_dag/util/command_patterns.py#L70) and [here](https://github.com/chanzuckerberg/idseq-dag/blob/b96a29efe73c6bc82ea7dbc8debb352f3a62dce0/idseq_dag/util/command_patterns.py#L212).

And we probably want to pass this session `boto` calls as well. The only place where it is being used idseq-dag is in this file.

Reply:
I agree it makes sense to pass this modified environment to subprocess calls directly instead of setting it across the idseq-dag process. I will implement that change as you propose.

We should not be passing this cached session to other boto3/botocore calls within this process. The only reason we need to pass it to the AWS CLI is that it is an independent process that does not cache credentials from run to run. Boto3/botocore in-process *do* cache their credentials, and we should not interfere with their default behavior, to minimize surprises/unexpected issues.

Reply:
@davidrissato I have implemented the changes you proposed.